src/rust-ffi/src/rangeproof.rs

```diff
@@ -0,0 +1,158 @@
+        point: Box::into_raw(Box::new(point)),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rangeproof_prove_multiple(
+    bp_gens: *const BulletproofGens,
+    pc_gens: *const PedersenGens,
+    v: *const u64,
+    v_blinding: *const Scalar, // this is a slice
+    n: u32,
+    label: *const u8,
+    label_len: i32,
+    amount: i32,
+) -> RangeProofWithCommit {
+    let n = n as usize;
+    let bp_gens = deref!(bp_gens);
+    let pc_gens = deref!(pc_gens);
+    let label = unsafe { slice::from_raw_parts(label, label_len as usize) };
+
+    let v_blinding = unsafe { slice::from_raw_parts(v_blinding, amount as usize) };
+    let v = unsafe { slice::from_raw_parts(v, amount as usize) };
+
+    let mut transcript = Transcript::new(label);
+    let proof = RangeProof::prove_multiple(bp_gens, pc_gens, &mut transcript, v, v_blinding, n);
+
+    let proof = proof.unwrap();
+    let (proof, points) = proof;
+    RangeProofWithCommit {
+        proof: Box::into_raw(Box::new(proof)),
+        point: points.as_ptr() as *mut _,
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rangeproof_verify_single(
+    proof: *mut RangeProof,
+    bp_gens: *const BulletproofGens,
+    pc_gens: *const PedersenGens,
+    commit: *const CompressedRistretto,
+    n: u32,
+    label: *const u8,
+    label_len: i32,
+) -> bool {
+    let proof = deref!(proof);
+    let n = n as usize;
+    let bp_gens = deref!(bp_gens);
+    let pc_gens = deref!(pc_gens);
+    let commit = deref!(commit);
+    let label = unsafe { slice::from_raw_parts(label, label_len as usize) };
+    let mut transcript = Transcript::new(label);
+
+    let res = proof.verify_single(bp_gens, pc_gens, &mut transcript, commit, n);
+    res.is_ok()
+}
+
+#[no_mangle]
+pub extern "C" fn rangeproof_verify_multiple(
+    proof: *mut RangeProof,
+    bp_gens: *const BulletproofGens,
+    pc_gens: *const PedersenGens,
+    commits: *const CompressedRistretto,
+    n: u32,
+    label: *const u8,
+    label_len: i32,
+    amount: u32,
+) -> bool {
+    let proof = deref!(proof);
+    let n = n as usize;
+    let bp_gens = deref!(bp_gens);
+    let pc_gens = deref!(pc_gens);
+    let commits = unsafe { slice::from_raw_parts(commits, amount as usize) };
+    let label = unsafe { slice::from_raw_parts(label, label_len as usize) };
+    let mut transcript = Transcript::new(label);
+
+    let res = proof.verify_multiple(bp_gens, pc_gens, &mut transcript, commits, n);
+    res.is_ok()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rangeproof_to_bytes(proof: *mut RangeProof) -> RawVec<u8> {
+    let mut proof = (*proof).to_bytes();
+    let raw = RawVec {
+        data: proof.as_mut_ptr(),
+        size: proof.len(),
+        cap: proof.capacity(),
+    };
+    std::mem::forget(proof);
+    raw
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rangeproof_from_bytes(bytes: *mut u8, len: u32) -> *const RangeProof {
+    let bytes = slice::from_raw_parts(bytes, len as usize);
+    let Ok(proof) = RangeProof::from_bytes(bytes) else {
```

Comment:
This is odd - the Ok(proof) function, what am I missing. It is not being used or does it transmute the proof? Maybe it is just me.